### ISSUE TITLE
[webview_flutter] At loading, scroll may not be at the top of the HTML page on some iOS webviews

### ISSUE THREAD

**Issue by stevemoreau** [flutter/flutter#47139]
## TL;DR:

Tab controller with several webviews (one per tab). The first tab renders a webview with scroll position at the top of the HTML content. The second tab renders the webview with a small offset on some iOS real devices (apparently jumping to 6% of the HTML page). This only occurs on real devices but not on simulators

![webview_scroll_ok](https://user-images.githubusercontent.com/16148036/70914912-35f8f200-2019-11ea-840c-47b848a0cf01.png) ![webview_scroll_ko](https://user-images.githubusercontent.com/16148036/70914928-3abda600-2019-11ea-8a58-8ff7271ca5a7.png)

## Steps to Reproduce

1. Clone https://github.com/stevemoreau/flutter_webview, build and run the flutter application *on real iOS device* (would work properly on simulators)
2. Checkout the `Tab 0`: the display should start at 0.50% (embedded html body text) and it is not possible to scroll up since we are at the top of the page
3. Go to `Tab 1`: the HTML rendered should not start at 0.50% as for `Tab 0` (6.00% in my case). It is possible to scroll up but by default the current scroll position is not at the top the HTML page

As a workaround, I tried injecting javascript to force the scroll to the top after loading (via [window.scrollTo](https://developer.mozilla.org/en-US/docs/Web/API/Window/scrollTo)) but it didn't work.

**Target Platform:** iOS **real** devices
**Target OS version: iOS 13.2.3**
**Devices: iPhone X**
**Flutter configuration:** webview_flutter: ^0.3.18+1

## Logs

$ flutter run --verbose
see [webview.txt](https://github.com/flutter/flutter/files/3968824/webview.txt)

```
$ flutter analyze
Analyzing flutter_webview_scroll_middlepage...                          

   info • 'ancestorInheritedElementForWidgetOfExactType' is deprecated and shouldn't be used. Use getElementForInheritedWidgetOfExactType instead. This feature was deprecated after v1.12.1. • lib/utils/i18n/bloc_provider_new.dart:25:17 • deprecated_member_use
   info • This method overrides a method annotated as @mustCallSuper in 'AutomaticKeepAliveClientMixin', but doesn't invoke the overridden method • lib/views/MyWebView.dart:75:10 • must_call_super

2 issues found. (ran in 8.2s)
```

```
$ flutter doctor -v
[✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.15.1 19B88, locale fr-FR)
    • Flutter version 1.12.13+hotfix.5 at /Users/steve/sdk/flutter
    • Framework revision 27321ebbad (6 days ago), 2019-12-10 18:15:01 -0800
    • Engine revision 2994f7e1e6
    • Dart version 2.7.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/steve/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3, Build version 11C29
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 41.1.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] Connected device (2 available)
    • Android SDK built for x86 • emulator-5554                            • android-x86 • Android 7.0 (API 24) (emulator)
    • iPhone                    • 5c21e6117ed37334bb86a3667172e7317c1f609e • ios         • iOS 13.2.3

• No issues found!
```

**Comment by mathieusachet** [comment]
same problem here

**Comment by danagbemava-nc** [comment]
The issue is reproducible on `master 2.5.0-6.0.pre.40` and `stable 2.2.3` with `webview_flutter: ^2.0.10`. 

https://user-images.githubusercontent.com/88313112/128828880-81df9e61-6af2-4f69-a789-e53be995cd5b.mov


<details>
<summary>code sample</summary>

```dart

import 'package:flutter/material.dart';
import 'package:webview_flutter/webview_flutter.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {
  int _counter = 0;
  late TabController _controller;

  final url =
      "https://app-mobile-webview.mobizel.com/webview_flutter/html_with_scroll_position.html";

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  void initState() {
    super.initState();
    _controller = new TabController(length: 5, vsync: this, initialIndex: 0);
  }

  @override
  Widget build(BuildContext context) {
    return DefaultTabController(
      length: 5,
      initialIndex: _controller.index,
      child: Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
          bottom: TabBar(
            controller: _controller,
            onTap: (index) => setState(() => _controller.index = index),
            tabs: [
              Tab(text: 'Tab 1'),
              Tab(text: 'Tab 2'),
              Tab(text: 'Tab 3'),
              Tab(text: 'Tab 4'),
              Tab(text: 'Tab 5'),
            ],
          ),
        ),
        body: IndexedStack(children: webviews, index: _controller.index),
        floatingActionButton: FloatingActionButton(
          onPressed: _incrementCounter,
          tooltip: 'Increment',
          child: const Icon(Icons.add),
        ),
      ),
    );
  }

  List<WebView> get webviews => [
        WebView(key: UniqueKey(), initialUrl: url),
        WebView(key: UniqueKey(), initialUrl: url),
        WebView(key: UniqueKey(), initialUrl: url),
        WebView(key: UniqueKey(), initialUrl: url),
        WebView(key: UniqueKey(), initialUrl: url),
      ];
}

```
</details>

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 2.5.0-6.0.pre.40, on macOS 11.5.1 20G80 darwin-arm, locale
    en-GH)
    • Flutter version 2.5.0-6.0.pre.40 at /Users/nexus/dev/sdks/flutters
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7d76a827c0 (9 hours ago), 2021-08-09 15:21:17 -0700
    • Engine revision 4fef55db10
    • Dart version 2.14.0 (build 2.14.0-385.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.59.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.25.0

[✓] Connected device (3 available)
    • Nexus’ IPhone (mobile) • 00008020-001875E83A38002E            • ios            • iOS
      14.6 18F72
    • iPhone 12 (mobile)     • F2592D43-5FE3-46BB-8A2B-A38BA468DCE9 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • Chrome (web)           • chrome                               • web-javascript •
      Google Chrome 92.0.4515.131

• No issues found!

```

```

[✓] Flutter (Channel stable, 2.2.3, on macOS 11.5.1 20G80 darwin-arm, locale en-GH)
    • Flutter version 2.2.3 at /Users/nexus/dev/sdks/flutter
    • Framework revision f4abaa0735 (6 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.59.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.25.0

[✓] Connected device (3 available)
    • Nexus’ IPhone (mobile) • 00008020-001875E83A38002E            • ios            • iOS
      14.6
    • iPhone 12 (mobile)     • F2592D43-5FE3-46BB-8A2B-A38BA468DCE9 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • Chrome (web)           • chrome                               • web-javascript •
      Google Chrome 92.0.4515.131

• No issues found!
```
</details>

**Comment by TranVanDinh235** [comment]
Any updates on this?

**Comment by joshmossas** [comment]
If anyone else needs a workaround. I got around this by running the following `webViewController?.runJavascript("window.scrollTo(0, 0);")` after the page has been loaded. I also had to delay the javascript execution with a `Timer` to get it to work. 

It's kind of hacky but it works I suppose.

__Complete example:__
```dart

class MyWebView extends StatefulWidget {
  const MyWebView({super.key});

  @override
  State<StatefulWidget> createState() => _MyWebView();
}

class _MyWebView extends State<MyWebView> {
  WebViewController? webViewController;

  @override
  Widget build(BuildContext context) {
    return WebView(
      javascriptMode: JavascriptMode.unrestricted,
      onPageFinished: (url) async {
        Timer(const Duration(seconds: 1), () async {
          webViewController?.runJavascript("window.scrollTo(0, 0);");
        });
      },
      onWebViewCreated: (controller) {
        setState(() {
          webViewController = controller;
        });
      },
      initialUrl: "https://example.com",
    );
  }
}

```

**Comment by jasonflaherty** [comment]
window.scrollTo(0,0) works for me, but using any type of document.getElementById('ID').scrollIntoView() or the like doesn't seem to work. Is there documentation I am missing on how to run other JS like that? 

```
_controller?.runJavaScript(
                  'const element = document.getElementById("pagetitletop");element.scrollIntoView();');
```

**Comment by dam-ease** [comment]
I am unable to reproduce this issue using the latest version of the `webview_flutter` package and on the latest `stable` and `master` channels. This issue seems to have been fixed on the latest versions. @stevemoreau  Kindly confirm this issue no longer persists using the latest versions.

<details><summary>Code Sample Used</summary>
<p>

```dart
import 'package:flutter/material.dart';
import 'package:webview_flutter/webview_flutter.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {
  int counter = 0;
  late TabController _controller;
  late final WebViewController _webViewController;

  final url =
      "https://app-mobile-webview.mobizel.com/webview_flutter/html_with_scroll_position.html";

  void _incrementCounter() {
    setState(() {
      counter++;
    });
  }

  @override
  void initState() {
    super.initState();
    _controller = new TabController(length: 5, vsync: this, initialIndex: 0);
    _webViewController = WebViewController();
  }

  @override
  Widget build(BuildContext context) {
    return DefaultTabController(
      length: 5,
      initialIndex: _controller.index,
      child: Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
          bottom: TabBar(
            controller: _controller,
            onTap: (index) {
              setState(() {
                _controller.index = index;
              });
            },
            tabs: [
              Tab(text: 'Tab 1'),
              Tab(text: 'Tab 2'),
              Tab(text: 'Tab 3'),
              Tab(text: 'Tab 4'),
              Tab(text: 'Tab 5'),
            ],
          ),
        ),
        body: WebViewWidget(
            controller: _webViewController
              ..loadRequest(
                Uri.parse(url),
              )),
        floatingActionButton: FloatingActionButton(
          onPressed: _incrementCounter,
          tooltip: 'Increment',
          child: const Icon(Icons.add),
        ),
      ),
    );
  }
}

``` 

</p>
</details> 

Although I noticed that with the latest version of the `webview_flutter` package, using the `IndexedStack` widget to display one `WebViewWidget` at one time by its index, only the final index loads the Web Page. This also occurs in the case of loading different web pages.

<details><summary>Code Sample</summary>
<p>
```dart
import 'package:flutter/material.dart';
import 'package:webview_flutter/webview_flutter.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {
  int counter = 0;
  late TabController _controller;
  late final WebViewController _webViewController;

  final url =
      "https://app-mobile-webview.mobizel.com/webview_flutter/html_with_scroll_position.html";

  void _incrementCounter() {
    setState(() {
      counter++;
    });
  }

  @override
  void initState() {
    super.initState();
    _controller = new TabController(length: 5, vsync: this, initialIndex: 0);
    _webViewController = WebViewController()
      ..loadRequest(
        Uri.parse(url),
      );
  }

  @override
  Widget build(BuildContext context) {
    return DefaultTabController(
      length: 5,
      initialIndex: _controller.index,
      child: Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
          bottom: TabBar(
            controller: _controller,
            onTap: (index) {
              setState(() {
                _controller.index = index;
              });
            },
            tabs: [
              Tab(text: 'Tab 1'),
              Tab(text: 'Tab 2'),
              Tab(text: 'Tab 3'),
              Tab(text: 'Tab 4'),
              Tab(text: 'Tab 5'),
            ],
          ),
        ),
        body: IndexedStack(index: _controller.index, children: webviews),
        floatingActionButton: FloatingActionButton(
          onPressed: _incrementCounter,
          tooltip: 'Increment',
          child: const Icon(Icons.add),
        ),
      ),
    );
  }

  List<Widget> get webviews => [
        WebViewWidget(controller: _webViewController),
        WebViewWidget(controller: _webViewController),
        WebViewWidget(controller: _webViewController),
        WebViewWidget(controller: _webViewController),
        WebViewWidget(controller: _webViewController),
      ];
}

``` 
</p>
</details>


<details> <summary> flutter doctor -v </summary>

```bash
[✓] Flutter (Channel master, 3.11.0-2.0.pre.55, on macOS 13.3 22E252 darwin-arm64, locale en-NG)
    • Flutter version 3.11.0-2.0.pre.55 on channel master at /Users/dammya/fvm/versions/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 835b892d7f (2 weeks ago), 2023-05-07 01:30:22 +0530
    • Engine revision 23f730efbf
    • Dart version 3.1.0 (build 3.1.0-83.0.dev)
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/dammya/Library/Android/sdk
    • Platform android-33, build-tools 32.1.0-rc1
    • ANDROID_HOME = /Users/dammya/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2022.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.78.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] VS Code (version 1.78.0)
    • VS Code at /Users/dammya/.Trash/Visual Studio Code 20.07.05.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (5 available)
    • sdk gphone arm64 (mobile)  • emulator-5554                            • android-arm64  • Android 11 (API 30) (emulator)
    • Damilola’s iPhone (mobile) • 71c843fa5f3dec767999f36224653c0faa78660e • ios            • iOS 16.4 20E247
    • iPhone 14 Pro Max (mobile) • D7F3302D-D66F-4B89-B8CF-DBEE1CDB5449     • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • macOS (desktop)            • macos                                    • darwin-arm64   • macOS 13.3 22E252 darwin-arm64
    • Chrome (web)               • chrome                                   • web-javascript • Google Chrome 113.0.5672.126

[✓] Network resources
    • All expected network resources are available.

• No issues found!

```
```bash
[✓] Flutter (Channel stable, 3.10.1, on macOS 13.3 22E252 darwin-arm64, locale
    en-NG)
    • Flutter version 3.10.1 on channel stable at /Users/dammya/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d3d8effc68 (5 days ago), 2023-05-16 17:59:05 -0700
    • Engine revision b4fb11214d
    • Dart version 3.0.1
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version
    32.1.0-rc1)
    • Android SDK at /Users/dammya/Library/Android/sdk
    • Platform android-33, build-tools 32.1.0-rc1
    • ANDROID_HOME = /Users/dammya/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2022.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.78.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] VS Code (version 1.78.0)
    • VS Code at /Users/dammya/.Trash/Visual Studio Code 20.07.05.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (5 available)
    • sdk gphone arm64 (mobile)  • emulator-5554                            •
      android-arm64  • Android 11 (API 30) (emulator)
    • Damilola’s iPhone (mobile) • 71c843fa5f3dec767999f36224653c0faa78660e •
      ios            • iOS 16.4 20E247
    • iPhone 14 Pro Max (mobile) • D7F3302D-D66F-4B89-B8CF-DBEE1CDB5449     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-arm64   • macOS 13.3 22E252 darwin-arm64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 113.0.5672.126

[✓] Network resources
    • All expected network resources are available.

• No issues found!

```
</details>
